src/charm.py

```diff
@@ -9,93 +9,306 @@
-
         self.framework.observe(self.on.get_credentials_action, self._on_get_credentials_action)
-        self.framework.observe(self.on.start, self._on_config_changed)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on[PEER].relation_changed, self._on_peer_relation_changed)
 
-        # MySQL
-        self.mysql = DatabaseRequires(self, relation_name="mysql", database_name=self.database)
-        self.framework.observe(self.mysql.on.database_created, self._on_database_created)
+        # Databases: MySQL, PostgreSQL and MongoDB
+        self.databases: Dict[str, DatabaseRequires] = {
+            name: self._setup_database_requirer(name) for name in DATABASES
+        }
 
-        # PostgreSQL
-        self.postgresql = DatabaseRequires(
-            self, relation_name="postgresql", database_name=self.database
+        # Kafka
+        self.kafka = KafkaRequires(
+            self,
+            relation_name=KAFKA,
+            topic=self.topic_name or "",
+            extra_user_roles=self.extra_user_roles or "",
         )
-        self.framework.observe(self.postgresql.on.database_created, self._on_database_created)
+        self.framework.observe(self.kafka.on.topic_created, self._on_topic_created)
+        self.framework.observe(self.on[KAFKA].relation_broken, self._on_relation_broken)
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle relation broken event."""
+        if not self.unit.is_leader:
+            return
+        # update peer databag to trigger the charm status update
+        self._update_active_relations(event, "broken")
 
-        # MongoDB
-        self.mongodb = DatabaseRequires(self, relation_name="mongodb", database_name=self.database)
-        self.framework.observe(self.mongodb.on.database_created, self._on_database_created)
+    def get_status(self) -> StatusBase:
+        """Return the current application status."""
+        if not self.topic_name and not self.database_name:
+            return BlockedStatus("Please specify either topic or database name")
 
-        self._on_config_changed(None)
+        if not self.is_database_related and not self.is_kafka_related:
+            return BlockedStatus("Please relate the data-integrator with the desired product")
+
+        if self.is_kafka_related and self.topic_active != self.topic_name:
+            logger.error("Trying to change Kafka configuration for existing relation")
+            return BlockedStatus("To change topic, please remove relation and add it again")
+
+        if self.is_database_related and any(
+            [database != self.database_name for database in self.databases_active.values()]
+        ):
+            logger.error("Trying to change database-name configuration for existing relation")
+            return BlockedStatus(
+                "To change database name, please remove relation and add it again"
+            )
```

Comment:
**suggestion:** Maybe it would be useful to loop over `self.databases_active.values()` at the top level, so that you can can extract `database-name`, and add that to the logging strings.

Reply:
+1 using walrus if the code does not become unreadable. @welpaolo trust your judgement. If it is too unreadable in your opinion, don't use walrus